fsfreeze.go

```diff
@@ -0,0 +1,150 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+)
+
+// from linux/fs.h
+const FIFREEZE = 0xC0045877
+const FITHAW = 0xC0045878
+
+const guestFsfreezeStatusThawed = "thawed"
+const guestFsfreezeStatusFrozen = "frozen"
+
+// Describes the system freeze status.
+var guestFsfreezeStatus string
+
+type MEntry struct {
+	FSSpec string
+	FSFile string
+	FSType string
+}
+
+// fsfreezeFreeze freezes the freezable mounted filesystems.
+func fsfreezeFreeze() (int, error) {
+	frozenFSCount := 0
+	m, err := getMountPoints()
+	if err != nil {
+		return 0, fmt.Errorf("error while getting mountpoints to fsfreeze-freeze: %v", err)
+	}
+
+	for _, mp := range m {
+		fs, err := os.Open(mp.FSFile)
+		if err != nil {
+			return 0, fmt.Errorf("error while opening mount path for fsfreeze-freeze: %v", err)
+		}
+
+		if err := ioctl(fs.Fd(), FIFREEZE, 0); err != nil &&
+			err.(*os.SyscallError).Err.(syscall.Errno) != syscall.EOPNOTSUPP &&
+			err.(*os.SyscallError).Err.(syscall.Errno) != syscall.EBUSY {
+			return 0, fmt.Errorf("error while executing fsfreeze-freeze on mount path %q: %v", mp.FSFile, err)
+		}
+
+		fs.Close()
```

Comment:
Here and below, you’re leaking file handles in the error case. Maybe use `defer fs.Close()` above instead? Having multiple file descriptors open while the loop is running shouldn’t be a problem.

Reply:
Ah yes, true, fixed.